### PR TITLE
Detect `Thread.Sleep` and Win32 `Sleep/SleepEx` as blocking objects 

### PR DESCRIPTION
**Pull request by Pavel-Durov**
Detecting `Thread.Sleep` and Win32 `Sleep/SleepEx` as blocking objects by inspecting for NtDelayExecution function. Same flow  as we do with CriticalSections

### CODE REVIEW

File: msos/StackWalker.cs [hunk 1]
```diff
@@ -20,29 +20,47 @@ public StackWalkerStrategy_x86(ClrRuntime runtime) : base(runtime)
...
         }
 
+
+        internal override UnifiedBlockingObject GetNtDelayExecutionBlockingObject(UnifiedStackFrame frame)
+        {
+            var parameters = GetParameters(frame, NTDELAY_EXECUTION_FUNCTION_PARAM_COUNT);
+            var largeIntegerAddress = ConvertToAddress(parameters[1]);
+
+            var largeInt = ReadStructureFromAddress<LARGE_INTEGER>(largeIntegerAddress);
+            var awaitMs = ((-1) * largeInt.QuadPart) / 10000;
```
Comment (goldshtn) [inline msos/StackWalker.cs]: `((-1) * something)` is really a very verbose way of saying `-something`.

File: msos/StackWalker.cs [hunk 2]
```diff
@@ -20,29 +20,47 @@ public StackWalkerStrategy_x86(ClrRuntime runtime) : base(runtime)
...
+        {
+            T result = default(T);
+
+            byte[] buffer = new byte[Marshal.SizeOf(typeof(T))];
             int read;
 
-            if (!_runtime.ReadMemory(criticalSectionAddress, buffer, buffer.Length, out read) || read != buffer.Length)
-                throw new Exception($"Error reading critical section data from address: {criticalSectionAddress}");
+            if (!_runtime.ReadMemory(address, buffer, buffer.Length, out read) || read != buffer.Length)
+                throw new Exception($"Error reading structure data from address: {address}");
```
Comment (goldshtn) [inline msos/StackWalker.cs]: Would be nice to format the address as hex here.

File: msos/WaitChains.cs [hunk 3]
```diff
@@ -108,7 +108,11 @@ private void DisplayChainForThread(UnifiedThread unifiedThread, int depth, HashS
         {
             foreach (var blockingObject in unifiedThread.BlockingObjects)
             {
-                _context.Write("{0}| {1} ", new string(' ', (depth + 1) * 2), blockingObject.Reason);
+                string reason = blockingObject.Type == UnifiedBlockingType.ThreadSleep
+                                ? $"ThreadSleep: {blockingObject.SleepWait.Value}"
+                                : blockingObject.Reason.ToString();
```
Comment (goldshtn) [inline msos/WaitChains.cs]: I don't understand why this special treatment is needed here. Why can't the blocking object itself do its own formatting? This kind of if-else logic is really unclean.

**Review by goldshtn — CHANGES_REQUESTED** [review]

File: msos/StackWalker.cs (hunk 1 above)
Comment (Pavel-Durov) [reply msos/StackWalker.cs]: changed

File: msos/StackWalker.cs (hunk 2 above)
Comment (Pavel-Durov) [reply msos/StackWalker.cs]: done

File: msos/WaitChains.cs (hunk 3 above)
Comment (Pavel-Durov) [reply msos/WaitChains.cs]: I added ReasonDescription property in UnifiedBlockingObject class, previously only Reason property (which is an UnifiedBlockingReason enum) was printed.

**Comment by Pavel-Durov** [comment]
@goldshtn Is there anything else I need to address in this PR? I think I went over all of the requested changes :)

File: msos/MixedStack.cs [hunk 4]
```diff
@@ -517,6 +517,13 @@ public UnifiedBlockingObject(ulong handle, string objectName, string objectType)
             Reason = ConvertToUnified(objectType);
         }
 
+        public UnifiedBlockingObject(long msTimeout) : this(BlockingObjectOrigin.StackWalker)
+        {
+            Type = UnifiedBlockingType.ThreadSleep;
+            Reason = UnifiedBlockingReason.ThreadWait;
+            ReasonDescription = $"MsTimeout : {msTimeout}";
```
Comment (goldshtn) [inline msos/MixedStack.cs]: Please change to read "Sleep with timeout: 766ms" if there is a timeout, or to "Sleep with no timeout" if the timeout is infinite.

File: msos/NativeStructs.cs [hunk 5]
```diff
@@ -51,6 +51,17 @@ public struct _WAITCHAIN_NODE_INFO_THREAD_OBJECT
 
         #region NtDll
 
+        [StructLayout(LayoutKind.Explicit, Size = 8)]
+        public struct LARGE_INTEGER
```
Comment (goldshtn) [inline msos/NativeStructs.cs]: Why is this needed? It's basically `Int64`.

**Review by goldshtn — CHANGES_REQUESTED** [review]
Just a couple of minor changes and this is good to go.

**Pull request merged**